### PR TITLE
templates: add toleration to control-plane components

### PR DESCRIPTION
**Pull request by squat**
This PR adds taints and tolerations to the bootkube templates to ensure that node workloads do not get scheduled on master nodes.

cc @aaronlevy @yifan-gu

### CODE REVIEW

File: pkg/asset/internal/templates.go [hunk 1]
```diff
@@ -599,6 +626,14 @@ spec:
         - name: etc-kubernetes
           mountPath: /etc/kubernetes
           readOnly: true
+      hostNetwork: true
+      tolerations:
+      - key: CriticalAddonsOnly
```
Comment (squat) [inline pkg/asset/internal/templates.go]: This toleration was missing entirely from this critical pod, meaning it wouldn't get scheduled when it was dire

File: hack/multi-node/bootkube-up [hunk 2]
```diff
@@ -22,7 +22,7 @@ if [ ! -d "cluster" ]; then
   ../../_output/bin/${local_os}/bootkube render --asset-dir=cluster --api-servers=https://172.17.4.101:443 ${etcd_render_flags}
   cp user-data.sample cluster/user-data-worker
   cp user-data.sample cluster/user-data-controller
-  sed -i.bak -e '/--node-labels=node-role.kubernetes.io\/master/d' cluster/user-data-worker
+  sed -i.bak -e '/node-role.kubernetes.io\/master/d' cluster/user-data-worker
```
Comment (squat) [inline hack/multi-node/bootkube-up]: Modifies the regex to ensure sed removes both the lines with `node-labels` and `register-with-taints`

**Comment by squat** [comment]
failed on
```
Total time extracting prefetched containers to disk: 37.369150289s
--- FAIL: bootkube.selfetcd.smoke (179.01s)
        run.go:148: creating cluster: failed to request new GCE instance: googleapi: Error 503: Code: '3530860079175192029', backendError
```

**Comment by squat** [comment]
rktbot run tests

File: pkg/asset/internal/templates.go [hunk 3]
```diff
@@ -431,6 +445,13 @@ spec:
         - name: ssl-host
           mountPath: /etc/ssl/certs
           readOnly: true
+      nodeSelector:
+        node-role.kubernetes.io/master: ""
+      tolerations:
```
Comment (yifan-gu) [inline pkg/asset/internal/templates.go]: This manifest already deinifes a `tolerations` field, can you merge them? @squat

File: pkg/asset/internal/templates.go [hunk 4]
```diff
@@ -540,7 +559,13 @@ spec:
             port: 10251  # Note: Using default port. Update if --port option is set differently.
           initialDelaySeconds: 15
           timeoutSeconds: 15
-
+      nodeSelector:
+        node-role.kubernetes.io/master: ""
+      tolerations:
```
Comment (yifan-gu) [inline pkg/asset/internal/templates.go]: ditto

File: pkg/asset/internal/templates.go [hunk 5]
```diff
@@ -147,12 +152,6 @@ spec:
         checkpointer.alpha.coreos.com/checkpoint: "true"
         scheduler.alpha.kubernetes.io/critical-pod: ''
     spec:
-      tolerations:
-      - key: "CriticalAddonsOnly"
```
Comment (yifan-gu) [inline pkg/asset/internal/templates.go]: Why this is removed from kube-apiserver? Or is this a previous fault?

**Comment by yifan-gu** [comment]
Some small nits.

File: pkg/asset/internal/templates.go (hunk 3 above)
Comment (squat) [inline pkg/asset/internal/templates.go]: Good catch. fixed

File: pkg/asset/internal/templates.go (hunk 4 above)
Comment (squat) [inline pkg/asset/internal/templates.go]: fixed

File: pkg/asset/internal/templates.go (hunk 5 above)
Comment (squat) [inline pkg/asset/internal/templates.go]: it was accidentaly removed while adding other toleration. fixed

**Comment by squat** [comment]
@yifan-gu thanks for the review. I addressed the nits and pushed the fixes

**Comment by yifan-gu** [comment]
LGTM

File: pkg/asset/internal/templates.go [hunk 6]
```diff
@@ -774,6 +795,18 @@ spec:
             memory: 20Mi
             cpu: 10m
       dnsPolicy: Default  # Don't use cluster DNS.
+      tolerations:
```
Comment (Quentin-M) [inline pkg/asset/internal/templates.go]: no nodeSelector in that one?

Comment (yifan-gu) [inline pkg/asset/internal/templates.go]: Good question, @aaronlevy

Comment (yifan-gu) [inline pkg/asset/internal/templates.go]: I don't see why not put it on master nodes.

Comment (aaronlevy) [inline pkg/asset/internal/templates.go]: Kind of up to us -- I don't have super strong feelings here.

Comment (yifan-gu) [inline pkg/asset/internal/templates.go]: I would prefer to put them on master nodes as DNS pod is virtually part of the control plane, it's almost always turned on by default.

Comment (aaronlevy) [inline pkg/asset/internal/templates.go]: wfm

Comment (squat) [inline pkg/asset/internal/templates.go]: added the nodeselector and pushed the commit

**Comment by yifan-gu** [comment]
LGTM

**Comment by Quentin-M** [comment]
@squat Trying to use your PR in Tectonic (v1.6.1_coreos.0) ... getting:

```
Error: invalid argument "node-role.kubernetes.io/master:NoSchedule" for --register-with-taints=node-role.kubernetes.io/master:NoSchedule: invalid taint spec: node-role.kubernetes.io/master:NoSchedule
```

**Comment by Quentin-M** [comment]
It has to be ```--register-with-taints=node-role.kubernetes.io/master=:NoSchedule```. Can you follow up?

**Comment by xiang90** [comment]
I do not feel we should get this merged. Not all master components (etcd) now tolerate this taints. We should make sure everything can tolerate taints before moving forward with this solution.

also i do not think the test should ever pass. either we taints incorrectly or we use an old version of k8s for testing.

can we revert this back?

**Comment by hongchaodeng** [comment]
Any testing on this?
IIUC, with this change, the testing still doesn't set any taints on node in testing. @pbx0 to confirm?
If so, we should revert this until testing is covered.

**Comment by aaronlevy** [comment]
The tests pass due to a bug in flannel which strips the taints off the node objects (https://github.com/coreos/flannel/issues/667). Once we merge a fix for that (https://github.com/kubernetes-incubator/bootkube/issues/454) self-hosted etcd will stop working until the pods tolerate the taint.

Rather than reverting this -- it seems like adding the toleration should be a relatively trivial change -- can we just get that merged?

**Comment by aaronlevy** [comment]
All other master components do already tolerate this taint - we should just need etcd pods as final step

**Comment by xiang90** [comment]
@aaronlevy We should never merge change that breaks test. If it breaks existing stuff, we need to fix first, then merge. It should never go the other way around. I still think we should revert the change until we figure everything out.

**Pull request merged**